generators/entity-server/templates/src/main/java/package/domain/Entity.java.ejs

```diff
@@ -700,49 +721,56 @@ relationships.forEach((relationship, idx) => {
             return false;
         }
         <%_ if (!embedded) { _%>
-        return id != null && id.equals(((<%= asEntity(entityClass) %>) o).id);
+        return <%= primaryKeyName %> != null && <%= primaryKeyName %>.equals(((<%= asEntity(entityClass) %>) o).<%= primaryKeyName %>);
         <%_ } else { _%>
         return false;
         <%_ } _%>
     }
 
     @Override
     public int hashCode() {
-        return 31;
+        return Objects.hashCode(<%= primaryKeyName %>);
```

Comment:
Should revert. See last PR.

Reply:
didn't find anything can you lease explain so I add a comment for why?

Reply:
PR https://github.com/jhipster/generator-jhipster/pull/13097, if you see the bug there is explanations of why hashCode() should be constant for jpa.
But basically in some cases, when generating the id after it's already is a Set, the hashCode changes, and the entity is lost causing dataloss.

Reply:
I just did some reading, and there is a real problem I didn't know about. basically for anyone else having the same hashcode for all objects force collections to call equals (that is always false if the id is not yet set under circumvents the use of buckets in Hashset/Hashmap) dropping performance but ensuring that checking if an object is in a collection always works

Reply:
Since this a problem only when the id is generated by the database, (not a foreign key...) I will use 31 only in that case.

Reply:
It’s not related to database, but related to lazily created ids. hashCode needs to be constant.
A composite id can have a generated field with a custom sequence or based on row number. So not safe.